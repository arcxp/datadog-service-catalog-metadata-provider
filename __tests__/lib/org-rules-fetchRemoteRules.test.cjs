// Seed these for GitHub's toolkit
const path = require('path')

const YAML = require('yaml')

// Pulling this in here activates the mocking of the github module
const github = require('@actions/github')

// Need to use inputs for some of our parameters
const core = require('@actions/core')

const testLocallyOnly = require('../test-locally-only')

const {
  applyOrgRules,
  _test: {
    fetchRemoteRules,
    ghHandle,
    determineApplicabilityOfRule,
    determineRuleCompliance,
  },
} = require('../../lib/org-rules')

process.env.GITHUB_EVENT_PATH = path.join(
  __dirname,
  '../data/github-context-payload.json',
)
process.env.GITHUB_REPOSITORY =
  'arcxp/datadog-service-catalog-metadata-provider'

describe('org-rules.js Org Rules File acquisition', () => {
  let gh = undefined

  beforeAll(async () => {
    gh = await ghHandle()
  })

  // Reset inputs before each test
  beforeEach(() => {
    core.__resetInputsObject()
  })

  test('#fetchRemoteRules() - default case: no org rules file', async () => {
    const result = await fetchRemoteRules(
      gh,
      'org-rule-test-file-not-found.yml',
    )
    expect(result).toMatchObject({
      org: 'arcxp',
      rules: expect.arrayContaining([]),
    })
  })

  test('#fetchRemoteRules() - default case: has org rules file', async () => {
    const result = await fetchRemoteRules(gh)
    expect(result).toMatchObject({
      org: 'arcxp',
      rules: expect.arrayContaining([]),
    })
    // Yay for snapshot testing!
    expect(result).toMatchSnapshot()
  })

  test('#fetchRemoteRules() - exception case: failed to parse', async () => {
    const orig_core_setFailed = core.setFailed
    core.setFailed = jest.fn()
    const result = await fetchRemoteRules(
      gh,
      'broken-service-catalog-rules.yml',
    )
    expect(core.setFailed).toBeCalledTimes(1)
    expect(core.setFailed).toBeCalledWith(
      'Org Rules File "broken-service-catalog-rules.yml" failed to parse.',
    )
    expect(result).toBeUndefined()
    core.setFailed = orig_core_setFailed
  })

  testLocallyOnly('#fetchRemoteRules() - no discernable org', async () => {
    const old_gh_repo = process.env.GITHUB_REPOSITORY
    delete process.env.GITHUB_REPOSITORY
    core.setFailed = jest.fn()
    const result = await fetchRemoteRules()
    expect(core.setFailed).toBeCalledTimes(1)
    expect(core.setFailed).toHaveBeenCalledWith(
      'This GitHub Actions environment does not have a valid context.',
    )
    expect(result).toMatchObject({ org: 'UNKNOWN', rules: [] })

    process.env.GITHUB_REPOSITORY = old_gh_repo
  })

  afterEach(() => {
    jest.clearAllMocks()
  })
})
