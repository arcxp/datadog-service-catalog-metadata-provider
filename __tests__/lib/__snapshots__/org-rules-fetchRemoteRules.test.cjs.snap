// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org-rules.js Org Rules File acquisition #fetchRemoteRules() - default case: has org rules file 1`] = `
{
  "org": "arcxp",
  "rules": [
    {
      "name": "Simple Requirements",
      "requirements": {
        "tags": {
          "internet_accessible": "ANY",
        },
      },
      "selection": "all",
    },
    {
      "name": "Prod needs a runbook",
      "requirements": {
        "links": {
          "type": "runbook",
        },
      },
      "selection": {
        "tags": {
          "internet_accessible": "yes",
        },
      },
    },
  ],
}
`;
